### PR TITLE
WIP feat(search): default search hooks have been moved to core

### LINKED ISSUE #7392: Move search hooks to core
As noted in #7384, I would love to have an option of using search hooks without having the search plugin enabled. There are cases, when the search results pages are customized heavily, and heaving the search plugin with all of it's logic is unnecessary.

Additionally, it seems we are reinventing the wheel in many of the plugins, by rewriting where and join clauses for matching by name, title, description etc.

### PR DESCRIPTION
**Pull request by jdalsem**
BREAKING CHANGE:
Plugins relying on the search plugin should revisit the search related
code. Names of search hooks have been renamed.

Fixes #7392 
- [ ] Upgrade docs
- [x] Rework search internals (move to service)

### CODE REVIEW

**Comment by jdalsem** [comment]
This also makes changes like #8724 more logical

**Comment by jdalsem** [comment]
if this is in core... would it make sense to add a ``elgg_search`` function? instead of triggering a hook?

**Comment by hypeJunction** [comment]
I think there should only be one function that filters all wheres, joins and order by clauses. They go hand in hand and breaking them is what nakes the search plugins such a mess. Basically what I am saying a search hook shoukd just call get_clauses, search hook or something. I had a bit of a different vision for this but I guess we can refactor later.

**Comment by jdalsem** [comment]
Yes wanted to refactor this... That's why WIP.  Putting it out here for
input..

Op 10 okt. 2016 5:32 p.m. schreef "Ismayil Khayredinov" <
notifications@github.com>:

**Comment by hypeJunction** [comment]
Having a query builder would make life a lot easier.. sigh

**Comment by jdalsem** [comment]
Would like some eyes on the code @Elgg/core  before finishing up the documentation..

Main focus of this PR: bring search to core (centralized around `elgg_search`). Don't want anything output related in core... just the search queries. Output related stuff stays in the `search` plugin.

This allows search (queries/hooks) to be used without the need for the search plugin to be activated.

File: engine/lib/search.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
+<?php
+/**
+ * Basic Elgg search hooks.
```
Comment (jdalsem) [inline engine/lib/search.php]: need to update this description...

File: engine/lib/search.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+ * This is good for performance since search is slow and there are many pages all
+ * with the same content.
+ *
+ * @param string $hook Hook name
+ * @param string $type Hook type
+ * @param string $text robots.txt content for plugins
+ *
+ * @return string
+ */
+function elgg_search_exclude_robots($hook, $type, $text) {
```
Comment (jdalsem) [inline engine/lib/search.php]: should i make this a private function?

File: mod/search/views/default/search/object/comment/entity.php [hunk 3]
```diff
@@ -5,35 +5,35 @@
...
-$url = $entity->getURL();
-$title = "<a href=\"$url\">$title</a>";
+$title = elgg_view('output/url', [
+	'href' => $entity->getURL(),
+	'text' => $title_text,
+]);
 
 $description = $entity->getVolatileData('search_matched_description');
 
-$title = "<a href=\"$url\">$title</a>";
```
Comment (jdalsem) [inline mod/search/views/default/search/object/comment/entity.php]: why noone every reported this....

**Comment by hypeJunction** [comment]
I'd prefer we wait until we know what's happening with metadata/attributes. Utlimately, we may not even need a search service if we expand ege*

**Comment by jdalsem** [comment]
i prefer to keep search around as it serves a different purpose and is not bound to the metadata/attributes change...

simply said, this PR moves the hooks from the `search` plugin into core and adds a helper function around the hooks. This way it is much easier to replace search features. Currently you mostly build on top of the search plugin, but with this PR it allows to swap out the search plugin much more easily, without losing support of search capabilities in other plugins.

I developed multiple search plugins, and this approach would make life a lot easier.

**Comment by hypeJunction** [comment]
I am using this as basis for a new PR, don't put more effort into it for now

**Comment by hypeJunction** [comment]
Revamped search is here: #10688

**Comment by hypeJunction** [comment]
Let's focus on and finish #10688 instead

**Pull request closed without merging**